chain/chunks/src/lib.rs

```diff
@@ -998,78 +998,106 @@ impl ShardsManager {
-            // Create iterators which _might_ contain the requested parts.
-            let parts_iter = request.part_ords.iter().map(|ord| entry.parts.get(ord).cloned());
-            let receipts_iter = request
-                .tracking_shards
-                .iter()
-                .map(|shard_id| entry.receipts.get(shard_id).cloned());
-
-            // Pass iterators to function which will evaluate them. Since iterators are lazy
-            // we will clone as few elements as possible before realizing not all are present.
-            // In the case all are present, the response is sent.
-            return self.maybe_send_partial_encoded_chunk_response(
-                request.chunk_hash,
-                route_back,
-                parts_iter,
-                receipts_iter,
-            );
-        // If not in the cache then check the storage
+        let response = if let Some(entry) = self.encoded_chunks.get(&request.chunk_hash) {
+            Self::prepare_partial_encoded_chunk_response_from_cache(request, entry)
         } else if let Ok(partial_chunk) = chain_store.get_partial_chunk(&request.chunk_hash) {
-            // Index _references_ to the parts we know about by their `part_ord`. Since only
-            // references are used in this index, we will only clone the requested parts, not
-            // all of them.
-            let present_parts: HashMap<u64, _> =
-                partial_chunk.parts().iter().map(|part| (part.part_ord, part)).collect();
-            // Create an iterator which _might_ contain the request parts. Again, we are
-            // using the laziness of iterators for efficiency.
-            let parts_iter =
-                request.part_ords.iter().map(|ord| present_parts.get(ord).map(|x| *x).cloned());
-
-            // Same process for receipts as above for parts.
-            let present_receipts: HashMap<ShardId, _> = partial_chunk
-                .receipts()
-                .iter()
-                .map(|receipt| (receipt.1.to_shard_id, receipt))
-                .collect();
-            let receipts_iter = request
-                .tracking_shards
-                .iter()
-                .map(|shard_id| present_receipts.get(shard_id).map(|x| *x).cloned());
-
-            // Pass iterators to function, same as cache case.
-            return self.maybe_send_partial_encoded_chunk_response(
-                request.chunk_hash,
-                route_back,
-                parts_iter,
-                receipts_iter,
-            );
+            Self::prepare_partial_encoded_chunk_response_from_partial(request, partial_chunk)
+        } else {
+            None
         };
+
+        if let Some(response) = response {
+            self.peer_manager_adapter.do_send(PeerManagerMessageRequest::NetworkRequests(
+                NetworkRequests::PartialEncodedChunkResponse { route_back, response },
+            ))
+        }
```

Comment:
in separate PR:

maybe we should have some stats - on how many times we don't have the response (as normally this should not happen).